ts/nni_manager/common/gpu_scheduler/scheduler.ts

```diff
@@ -0,0 +1,246 @@
+
+    /**
+     *  Release a trial's allocated GPUs
+     **/
+    public async release(experimentId: string, trialId: string): Promise<void> {
+        this.releaseByFilter(trial => (trial.experimentId === experimentId && trial.trialId === trialId));
+    }
+
+    /**
+     *  Release all trials of an experiment.
+     *
+     *  Useful when the experiment is shutting down or has lost response.
+     **/
+    public async releaseAll(experimentId: string): Promise<void> {
+        logger.info('Release whole experiment', experimentId);
+        this.releaseByFilter(trial => (trial.experimentId === experimentId));
+    }
+
+    private updateGpus(info: GpuSystemInfo): void {
+        for (const gpu of info.gpus) {
+            const index = gpu.index;
+            this.gpus[index].coreUtil = gpu.gpuCoreUtilization ?? 0;
+            this.gpus[index].memUtil = gpu.gpuMemoryUtilization ?? 0;
+            this.gpus[index].active = false;
+            this.gpus[index].computeActive = false;
+        }
+
+        for (const proc of info.processes) {
+            const index = proc.gpuIndex;
+            this.gpus[index].active = true;
+            if (proc.type === 'compute') {
+                this.gpus[index].computeActive = true;
+            }
+        }
+    }
+
+    private sortGpus(restrict: ScheduleRestrictions): SchedulerGpuInfo[] {
+        let gpus = this.gpus.slice();  // copy for in-place sort
+        if (restrict.onlyAcceptIndices) {
+            gpus = gpus.filter(gpu => restrict.onlyAcceptIndices!.includes(gpu.index));
+        }
+        if (restrict.rejectActiveGpus) {
+            gpus = gpus.filter(gpu => !gpu.active);
+        }
+        if (restrict.rejectComputeActiveGpus) {
+            gpus = gpus.filter(gpu => !gpu.computeActive);
+        }
+
+        // prefer the gpu with lower theoretical utilization;
+        // then the gpu without competing processes;
+        // then the gpu with more free memory;
+        // and finally the gpu with lower cuda core load.
+        return gpus.sort((a, b) => {
+            if (a.util !== b.util) {
+                return a.util - b.util;
+            }
+            if (a.active !== b.active) {
+                return Number(a.active) - Number(b.active);
+            }
+            if (a.computeActive !== b.computeActive) {
+                return Number(a.computeActive) - Number(b.computeActive);
+            }
+            if (a.memUtil !== b.memUtil) {
+                return a.memUtil - b.memUtil;
+            }
+            if (a.coreUtil !== b.coreUtil) {
+                return a.coreUtil - b.coreUtil;
+            }
+            return a.index - b.index;
+        });
+    }
+
+    private releaseByFilter(filter: (trial: SchedulerTrialInfo) => boolean): void {
+        const trials = this.trials.filter(filter);
+        trials.forEach(trial => {
+            logger.debug(`Released ${trial.experimentId}/${trial.trialId}`);
+            this.gpus[trial.gpuIndex].util -= trial.util;
```

Comment:
`this.gpus` might not be correct when someone is sharing the machine with others.

Reply:
That's why it will be a daemon.